### ISSUE TITLE
Aruco Simulation gone wrong

### ISSUE THREAD

**Issue by vejc94** [vejc94/project_dynamic_marker#3]
The problem with the simulation was that for each iteration of the simulation, I added an error to the camera parameters of the aruco simulation.  I changed it in the last commit. Now the errors of the z coordinate have changed. I think even thought the errors shakes for near positions, the simulation behaves more as one would expect. 
![aruco_n](https://user-images.githubusercontent.com/45331296/51121717-5ff9d100-1818-11e9-95a5-3a27b6dd4f1b.png)
![aruco_i](https://user-images.githubusercontent.com/45331296/51121718-5ff9d100-1818-11e9-8fcc-921bf524db33.png)

**Comment by raultron** [comment]
OK so, in order to produce better results you should take several measurements in each of the positions and then calculate the average and standard deviation (similar to what I did on the points configuration paper). I am still wondering why only one marker is having that weird behavior, I think it is a simulation problem, did you try to put the camera not exactly on the center?

**Comment by raultron** [comment]
Also, I still don't understand why you can't detect the circles at more than 0.6 meters. Can you show a screenshot of how the marker looks at that distance? if you can detect 6 aruco markers at 2m you should be able to detect 4 single circles on the screen.

**Comment by vejc94** [comment]
I am really not sure why at that position the error behaves that way. I also believe it is because some error in the simulation. I had tried with an inclination of 10 degree in the y-axis and it behaves similar to the first plot.
![aruco_error_10deg](https://user-images.githubusercontent.com/45331296/51234761-f8559a00-196d-11e9-856d-5c0fc37144ac.png)

**Comment by vejc94** [comment]
the algorithm is able to detect circle at 2 meters. You can see the results of this with the plot of the fixed marker. 
![fixed_marker](https://user-images.githubusercontent.com/45331296/51234878-57b3aa00-196e-11e9-9186-9dad115c2a0c.png)
With the dynamic marker it depends on the radius of the displayed circle. We know the circles' projections on the camera should have a minimum size (R=50 or R=80). Taking that into account if the camera is to far away i can't display more than 4 circles with the desired size.
![marker_r50](https://user-images.githubusercontent.com/45331296/51235264-59ca3880-196f-11e9-916e-c1dfb597a0e6.png)

**Comment by raultron** [comment]
A compromise that you could do here is to lower the limit of the minimum circle size for bigger distances. So starting from a given distance, you display a static marker of only 4 circles at the size that they fit on the screen or, as you mentioned in the other issue, you can use a single circle with conics. Something that I don't understand is that you show a plot at the beginning where you get detection at 2m but the last two plots have a lower maximum distance of detection.

**Comment by vejc94** [comment]
>Something that I don't understand is that you show a plot at the beginning where you get detection at 2m but the last two plots have a lower maximum distance of detection.

the first plot is with a marker that has a fixed number of circles a fixed size such as in the image below.
![fixed marker](https://user-images.githubusercontent.com/45331296/51249749-88a6d580-1994-11e9-9e98-dfb7c66dcad9.png)

the other plots shows that with the marker i can't estimate pose up until that maximum distances.

**Comment by raultron** [comment]
OK, now I understand. But this means that it is indeed possible to detect circles at these distances.
Since you designed the dynamic marker with a "desired" constraint for the size of the circle on the image (based on the accuracy of circle center detection), and that is why eventually you run out of space and the only option is to display one circle.

So to summarize, you have two options:

1- You keep this "desired size constrain" until the moment where you have 4 circles on the screen and for bigger distances the marker remains static (you don't change the size anymore).

2- When you have to display less than 4 circles you use the conic approach of whycon, or directly whycon markers. This means that if the camera is coming close to the screen from far away you will se first only one Whycon marker, that reduces size until you have two whycon markers and so on, and then you switch to the regular circles.

**Comment by vejc94** [comment]
Exactly. I am going to try to estimate pose with conics (as mentioned in the other issue). Having done that I would like to compare the 2 options that you mentioned. In other words, find out if it is better two estimate pose from farther distances with 4 circles (solvePnP of opencv) or with conics.

**Comment by vejc94** [comment]
Yesterday I implemented the metrics of your paper and made sure that they work. Now I realized that i have some serious problem with the pose estimation of aruco. As you see the coordinate frame of aruco is in the lower left corner of the board and in ViSP framework the world coordinate is in the middle of the image. I am unable to change both coordinate frames so when I compare the pose estimation from aruco with the real one I get some nasty results. I am not sure if there is a method to change the estimated rotation and translation vectors of aruco so that I can compare them with the real ones
![pregunta](https://user-images.githubusercontent.com/45331296/51387766-8f198680-1b27-11e9-9f81-3e294ae8d070.png)

**Comment by raultron** [comment]
You can create a transformation matrix that changes coordinates in Aruco lower left coordinate system into your world coordinate system. This should be quite easy to implement because it is only a 180 degrees rotation around the Xw axis and a translation.

**Comment by vejc94** [comment]
I also thought that it is easy to implement but I am struggling with it. Here is the code:
`void tranformCoordinatesAruco(cv::Vec3d r_ar, cv::Vec3d t_ar, cv::Vec3d &t_w, cv::Vec3d &r_w, cv::Ptr<cv::aruco::GridBoard> board){
    //marker length parameters
    double length =  board->getMarkerLength();
    //double gap = board->getMarkerSeparation();



    //homogeneous representation
    cv::Vec4d r_h, t_h, r_new, t_new;
    r_h(0) = r_ar(0);r_h(1) = r_ar(1);r_h(2) = r_ar(2);r_h(3) = 0;
    t_h(0) = t_ar(0);t_h(1) = t_ar(1);t_h(2) = t_ar(2);t_h(3) = 0;

    //transformation matrix from aruco to world frame
    vpHomogeneousMatrix aMw = vpHomogeneousMatrix(2*length, length, 0, vpMath::rad(180),vpMath::rad(0),vpMath::rad(0));

    //transforming
    for(int i = 0; i < 3; i++){
        for(int j = 0; j < 3; j++){
            r_new(i) += aMw[i][j]*r_h[j];
            t_new(i) += aMw[i][j]*t_h[j];
        }
    }

//saving
    t_w(0) = t_new(0);t_w(1) = t_new(1);t_w(2) = t_new(2);
    r_w(0) = r_new(0);r_w(2) = r_new(2);r_w(2) = r_new(2);`
